ecl/hql/hqlgram2.cpp

```diff
@@ -10950,68 +10952,190 @@ inline bool containsToken(int first, const int * expected)
     return false;
 }
 
-inline void removeToken(int token, int * expected)
+static bool isIdentifier(int token)
 {
-    for (int *finger = expected;*finger; finger++)
+    switch(token)
     {
-        if (*finger == token)
-        {
-            *finger = ' ';
-            break;
-        }
+    case DATAROW_ID:
+    case DATASET_ID:
+    case DICTIONARY_ID:
+    case SCOPE_ID:
+    case VALUE_ID:
+    case VALUE_ID_REF:
+    case ACTION_ID:
+    case UNKNOWN_ID:
+    case ENUM_ID:
+    case RECORD_ID:
+    case ALIEN_ID:
+    case TRANSFORM_ID:
+    case PATTERN_ID:
+    case FEATURE_ID:
+    case EVENT_ID:
+    case LIST_DATASET_ID:
+        return true;
+    default:
+        return false;
+    }
+}
+
+static bool isFunction(int token)
+{
+    switch (token)
+    {
+    case DATAROW_FUNCTION:
+    case DATASET_FUNCTION:
+    case DICTIONARY_FUNCTION:
+    case VALUE_FUNCTION:
+    case ACTION_FUNCTION:
+    case PATTERN_FUNCTION:
+    case RECORD_FUNCTION:
+    case EVENT_FUNCTION:
+    case SCOPE_FUNCTION:
+    case TRANSFORM_FUNCTION:
+    case LIST_DATASET_FUNCTION:
+        return true;
+    default:
+        return false;
+    }
+}
+
+static bool isType(int token)
+{
+    switch (token)
+    {
+    case TYPE_ID:
+    case SET_TYPE_ID:
+    case PATTERN_TYPE_ID:
+    case DATASET_TYPE_ID:
+    case DICTIONARY_TYPE_ID:
+        return true;
+    default:
+        return false;
+    }
+}
+
+static bool isExpression(int token)
+{
+    switch (token)
+    {
+    case ABS:
+    case ACOS:
+    case ASIN:
+    case ATAN:
+    case ATAN2:
+    case AVE:
+    case COS:
+    case COSH:
+    case DIV:
+    case EXP:
+    case TOK_LOG:
+    case LN:
+    case POWER:
+    case SIN:
+    case SINH:
+    case SQRT:
+    case TAN:
+    case TANH:
```

Comment:
This list is a mixture of expressions and some infix operators (/,%) which I don't think should be included in this list.